# Functions II Homework

### Preparation

In [1]:
# importing packages used throughout
import pandas as pd
import numpy as np
import os
import random
import statsmodels.api as sm
from math import sqrt

# directory work
os.getcwd()
os.chdir(r"C:\Users\natal\OneDrive\Documents\GitHub\QMSS-GR5072-Spring2026-HWs\homework-4-main\Submissions\MASETTI-NATALIE") # working directory
path = r"C:\Users\natal\OneDrive\Documents\GitHub\QMSS-GR5072-Spring2026-HWs\homework-4-main\Submissions\MASETTI-NATALIE" # path to hw data

In [2]:
# Load data in
data = pd.read_csv(path + r"/Housing_prices_data.csv") 

### Question #1: Import the data as a pd.DataFrame object. Use one or more functions from the pandas library to split the data into three unique datasets. The first should include columns that are factors only (i.e., categorical data), the second should include columns that are numeric only, and the third should include columns with logical/boolean variables only.

In [3]:
# already imported df as pd.DataFrame

In [4]:
# will use filter to split into 3 new datasets
# first identify column types
pd.options.display.max_columns = None
data.head()

# Id is necessary for all

#Logical: CentralAir; PavedDrive

# Categorical: MSZoning; Street; Alley; LotShape; LandContour; Utilities; LotConfig; LandSlope; Neighborhood; Condition1; Condition2
## BldgType; HouseStyle; RoofStyle; RoofMatl; Exterior1st; Exterior2nd; MasVnrType; ExterQual; ExterCond; Foundation; BsmtQual; BsmtCond;
## BsmtExposure; BsmtFinType1; BsmtFinType2; Heating; HeatingQC; Electrical; KitchenQual; Functional; FireplaceQu; GarageType; 
## GarageFinish; GarageQual; GarageCond; PoolQC?; Fence; MiscFeature; SaleType; SaleCondition

# Numeric: LotFrontage; LotArea; OverallQual; OverallCond; YearBuilt; YearRemodAdd; MasVnrArea; BsmtFinSF1; BsmtFinSF2; BsmtUnfSF; TotalBsmtSF
## 1stFlrSF; 2ndFlrSF; LowQualFinSF; GrLivArea; BsmtFullBath; BsmtHalfBath; FullBath; HalfBath; BedroomAbvGr; KitchenAbvGr; TotRmsAbvGrd
## Fireplaces; GarageYrBlt; GarageCars; GarageArea; WoodDeckSF; OpenPorchSF; EnclosedPorch; 3SsnPorch; ScreenPorch; PoolArea; MiscVal;
## MoSol; YrSold; SalePrice; MSSubClass

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,Condition2,BldgType,HouseStyle,OverallQual,OverallCond,YearBuilt,YearRemodAdd,RoofStyle,RoofMatl,Exterior1st,Exterior2nd,MasVnrType,MasVnrArea,ExterQual,ExterCond,Foundation,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinSF1,BsmtFinType2,BsmtFinSF2,BsmtUnfSF,TotalBsmtSF,Heating,HeatingQC,CentralAir,Electrical,1stFlrSF,2ndFlrSF,LowQualFinSF,GrLivArea,BsmtFullBath,BsmtHalfBath,FullBath,HalfBath,BedroomAbvGr,KitchenAbvGr,KitchenQual,TotRmsAbvGrd,Functional,Fireplaces,FireplaceQu,GarageType,GarageYrBlt,GarageFinish,GarageCars,GarageArea,GarageQual,GarageCond,PavedDrive,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,Norm,1Fam,2Story,7,5,2003,2003,Gable,CompShg,VinylSd,VinylSd,BrkFace,196.0,Gd,TA,PConc,Gd,TA,No,GLQ,706,Unf,0,150,856,GasA,Ex,True,SBrkr,856,854,0,1710,1,0,2,1,3,1,Gd,8,Typ,0,NaN,Attchd,2003.0,RFn,2,548,TA,TA,Y,0,61,0,0,0,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,FR2,Gtl,Veenker,Feedr,Norm,1Fam,1Story,6,8,1976,1976,Gable,CompShg,MetalSd,MetalSd,NaN,0.0,TA,TA,CBlock,Gd,TA,Gd,ALQ,978,Unf,0,284,1262,GasA,Ex,True,SBrkr,1262,0,0,1262,0,1,2,0,3,1,TA,6,Typ,1,TA,Attchd,1976.0,RFn,2,460,TA,TA,Y,298,0,0,0,0,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,Norm,1Fam,2Story,7,5,2001,2002,Gable,CompShg,VinylSd,VinylSd,BrkFace,162.0,Gd,TA,PConc,Gd,TA,Mn,GLQ,486,Unf,0,434,920,GasA,Ex,True,SBrkr,920,866,0,1786,1,0,2,1,3,1,Gd,6,Typ,1,TA,Attchd,2001.0,RFn,2,608,TA,TA,Y,0,42,0,0,0,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,Corner,Gtl,Crawfor,Norm,Norm,1Fam,2Story,7,5,1915,1970,Gable,CompShg,Wd Sdng,Wd Shng,NaN,0.0,TA,TA,BrkTil,TA,Gd,No,ALQ,216,Unf,0,540,756,GasA,Gd,True,SBrkr,961,756,0,1717,1,0,1,0,3,1,Gd,7,Typ,1,Gd,Detchd,1998.0,Unf,3,642,TA,TA,Y,0,35,272,0,0,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,FR2,Gtl,NoRidge,Norm,Norm,1Fam,2Story,8,5,2000,2000,Gable,CompShg,VinylSd,VinylSd,BrkFace,350.0,Gd,TA,PConc,Gd,TA,Av,GLQ,655,Unf,0,490,1145,GasA,Ex,True,SBrkr,1145,1053,0,2198,1,0,2,1,4,1,Gd,9,Typ,1,TA,Attchd,2000.0,RFn,3,836,TA,TA,Y,192,84,0,0,0,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [5]:
log_only = data.loc[:,['Id', 'CentralAir', 'PavedDrive']] # boolean only columns df

In [6]:
num_only = data.loc[:, ['Id', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 
    'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 
    'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 
    'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 
    'MiscVal', 'MoSold', 'YrSold', 'SalePrice', 'MSSubClass']]
# numeric only columns df

In [7]:
cat_only = data.loc[:, ['Id', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 
    'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 
    'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 
    'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 
    'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType', 'SaleCondition']]
# categorical only columns df

### Question #2: Using the second dataset from question #1, for each numeric variable use the apply function from the pandas library to return the mean, median and standard deviation (SD) of all numeric variables.

In [8]:
# using num_only, calculate mean, median, and SD
num2 = num_only.apply([np.mean, np.median, np.std], axis = 0)
num2

,Id,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,BsmtFinSF2,BsmtUnfSF,TotalBsmtSF,1stFlrSF,2ndFlrSF,LowQualFinSF,GrLivArea,BsmtFullBath,BsmtHalfBath,FullBath,HalfBath,BedroomAbvGr,KitchenAbvGr,TotRmsAbvGrd,Fireplaces,GarageYrBlt,GarageCars,GarageArea,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,MiscVal,MoSold,YrSold,SalePrice,MSSubClass
mean,730.500000,70.049958,10516.828082,6.099315,5.575342,1971.267808,1984.865753,103.685262,443.639726,46.549315,567.240411,1057.429452,1162.626712,346.992466,5.844521,1515.463699,0.425342,0.057534,1.565068,0.382877,2.866438,1.046575,6.517808,0.613014,1978.506164,1.767123,472.980137,94.244521,46.660274,21.954110,3.409589,15.060959,2.758904,43.489041,6.321918,2007.815753,180921.195890,56.897260
median,730.500000,NaN,9478.500000,6.000000,5.000000,1973.000000,1994.000000,NaN,383.500000,0.000000,477.500000,991.500000,1087.000000,0.000000,0.000000,1464.000000,0.000000,0.000000,2.000000,0.000000,3.000000,1.000000,6.000000,1.000000,NaN,2.000000,480.000000,0.000000,25.000000,0.000000,0.000000,0.000000,0.000000,0.000000,6.000000,2008.000000,163000.000000,50.000000
std,421.465598,24.274639,9977.846105,1.382523,1.112418,30.192559,20.638335,181.003845,455.941866,161.264017,441.715605,438.555057,386.455322,436.378914,48.606427,525.300394,0.518733,0.238671,0.550727,0.502713,0.815499,0.220263,1.624837,0.644446,24.680771,0.747059,213.731608,125.295863,66.233333,61.098214,29.307289,55.738317,40.163545,495.953090,2.702700,1.327640,79415.291886,42.286082


### Question #3: Using the second dataset from question #1, create a new dataset that only includes the variable indicating sales price ("SalePrice"). Search for a categorical variable which has between 2-4 categories in the full dataset that you think may be related to the sales price of houses. Give the mean, median and SD for SalePrice based on the different groups of the categorical variables. Does the average price seem to vary by the different levels of the categorical factor you chose?

In [9]:
sales_price = num_only['SalePrice'] # new dataframe, just saleprice

In [10]:
cat_only.head()
# searching catergorical variables

,Id,MSZoning,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,Condition2,BldgType,HouseStyle,RoofStyle,RoofMatl,Exterior1st,Exterior2nd,MasVnrType,ExterQual,ExterCond,Foundation,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinType2,Heating,HeatingQC,Electrical,KitchenQual,Functional,FireplaceQu,GarageType,GarageFinish,GarageQual,GarageCond,PoolQC,Fence,MiscFeature,SaleType,SaleCondition
0,1,RL,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,Norm,1Fam,2Story,Gable,CompShg,VinylSd,VinylSd,BrkFace,Gd,TA,PConc,Gd,TA,No,GLQ,Unf,GasA,Ex,SBrkr,Gd,Typ,NaN,Attchd,RFn,TA,TA,NaN,NaN,NaN,WD,Normal
1,2,RL,Pave,NaN,Reg,Lvl,AllPub,FR2,Gtl,Veenker,Feedr,Norm,1Fam,1Story,Gable,CompShg,MetalSd,MetalSd,NaN,TA,TA,CBlock,Gd,TA,Gd,ALQ,Unf,GasA,Ex,SBrkr,TA,Typ,TA,Attchd,RFn,TA,TA,NaN,NaN,NaN,WD,Normal
2,3,RL,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,CollgCr,Norm,Norm,1Fam,2Story,Gable,CompShg,VinylSd,VinylSd,BrkFace,Gd,TA,PConc,Gd,TA,Mn,GLQ,Unf,GasA,Ex,SBrkr,Gd,Typ,TA,Attchd,RFn,TA,TA,NaN,NaN,NaN,WD,Normal
3,4,RL,Pave,NaN,IR1,Lvl,AllPub,Corner,Gtl,Crawfor,Norm,Norm,1Fam,2Story,Gable,CompShg,Wd Sdng,Wd Shng,NaN,TA,TA,BrkTil,TA,Gd,No,ALQ,Unf,GasA,Gd,SBrkr,Gd,Typ,Gd,Detchd,Unf,TA,TA,NaN,NaN,NaN,WD,Abnorml
4,5,RL,Pave,NaN,IR1,Lvl,AllPub,FR2,Gtl,NoRidge,Norm,Norm,1Fam,2Story,Gable,CompShg,VinylSd,VinylSd,BrkFace,Gd,TA,PConc,Gd,TA,Av,GLQ,Unf,GasA,Ex,SBrkr,Gd,Typ,TA,Attchd,RFn,TA,TA,NaN,NaN,NaN,WD,Normal


I think the quality of the exterior would be related to Sale Price.

In [11]:
# choosing columns
q3 = data.loc[:, ['SalePrice', 'ExterQual']] # exterqual has right # of levels

In [12]:
q3.groupby('ExterQual').describe() # mean, median and SD for SalePrice based on ExterQual

SalePrice                                                    \
              count           mean            std       min       25%   
ExterQual                                                               
Ex             52.0  367360.961538  116401.264200  160000.0  311404.0   
Fa             14.0   87985.214286   39826.918794   39300.0   60250.0   
Gd            488.0  231633.510246   71188.873899   52000.0  185000.0   
TA            906.0  144341.313466   42471.815703   34900.0  118589.5   

                                          
                50%        75%       max  
ExterQual                                 
Ex         364606.5  428788.50  755000.0  
Fa          82250.0  102000.00  200000.0  
Gd         220000.0  265984.25  745000.0  
TA         139450.0  165500.00  381000.0

*Does the average price seem to vary by the different levels of the categorical factor you chose?*

Yes - exterior qualities rated as "Ex," "Gd", or "TA" have much higher average sale prices (ranging from 144,341 to 367,360) than exterior qualities labeled as "Fa" (87,985). (I couldn't find exactly what these levels mean in the Kaggle data explanation).

### Question #4: Regress "SalePrice" on "LotArea", "OverallQual", "OverallCond" with the statsmodels library. Use the syntax below to replicate the output shown you are able to estimate this regression model.

In [13]:
# specify DVs and IVs
x = num_only[['LotArea', 'OverallQual', 'OverallCond']]
y = (num_only['SalePrice'])

# fit linear regression model
model = sm.OLS(y, x).fit()

# model summary
print(model.summary())

                                 OLS Regression Results                                
Dep. Variable:              SalePrice   R-squared (uncentered):                   0.940
Model:                            OLS   Adj. R-squared (uncentered):              0.939
Method:                 Least Squares   F-statistic:                              7546.
Date:                Thu, 26 Mar 2026   Prob (F-statistic):                        0.00
Time:                        12:23:47   Log-Likelihood:                         -17827.
No. Observations:                1460   AIC:                                  3.566e+04
Df Residuals:                    1457   BIC:                                  3.568e+04
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                  coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------

### Question #5: Write a for loop that loops through each level of the categorical factor you chose in Q3, where each iteration of the for loop subsets your data for one group of the categorical variable at a time, and then fits the same regression model specified above in Q4 k times, where k refers to the number of unique categories. On each iteration of your for loop, store the model.summary() information as a new element in a list object. Once your for loop is done running, print the list. There should be k regressions which you estimated. Does the effect of LotArea on SalePrice vary within the different levels of the categorical factor you chose? What is another, more common way to answer a question like this (i.e., a question about moderation or effect heterogeneity)? Hint: Don't forget to account for the standard error when comparing the regression slopes!

In [14]:
# categorical variable = ExterQual

classes = ['Ex', 'Fa', 'Gd', 'TA'] # classes of ExterQual

output = [] # set output list

for i in classes: # for each class
    data = data[data['ExterQual'] == i] # subset df for that class
    x = data[['LotArea', 'OverallQual', 'OverallCond']] 
    y = (data['SalePrice'])
    model = sm.OLS(y, x).fit()
    output.append((model.summary()))
    data = pd.read_csv(path + r"/Housing_prices_data.csv") 
    print(output)

[<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
Dep. Variable:              SalePrice   R-squared (uncentered):                   0.930
Model:                            OLS   Adj. R-squared (uncentered):              0.925
Method:                 Least Squares   F-statistic:                              216.3
Date:                Thu, 26 Mar 2026   Prob (F-statistic):                    3.00e-28
Time:                        12:23:47   Log-Likelihood:                         -673.50
No. Observations:                  52   AIC:                                      1353.
Df Residuals:                      49   BIC:                                      1359.
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                  coef    std err          t      P>|t|      [0.025    

Yes, the effect of LotArea on SalePrice differs within the different levels of ExterQual. The effect of LotArea on SalePrice is only significant for 'Gd' and 'TA' ExterQuals. The R-squared is about the same for each ExterQual, but, again, the "Gd" and "TA" ExterQuals have the most variance in SalePrice captured in their models. Intuitively, the standard errors for the 'Gd' and 'TA' ExterQuals are also much lower (<1) than the standard errors for the other two ExterQuals, indicating more precision. The LotAreas all caused increases in SalePrice for each ExterQual except "Ex," where it caused a slight decrease. 

To test the effect more simply, I would add an interaction term to the OLS model so that it subsets by ExterQual class in the model itself.

### Question #6: Using the first dataset from question #1, for each categorical variable use a for loop to print a frequency and relative frequency table for every variable. To make your output readable, for each variable you should print the following in this order, placing the variable name where you see blanks below.


"Frequency table for ______ ============"

Display the frequency table

Skip a line, and print "Relative frequency table for ______ ============"

Skip two lines before printing the data for the next categorical variable


In [15]:
#for i in cat_only.columns:
#    print(f'Frequency table for: {cat_only[i]}')
#    print(cat_only[i].value_counts())
#    print('\n')
#    print(f'Relative frequency table for: {i}')
#    print(round(cat_only[i].value_counts(normalize=True),2))
#    print('\n')
#    print('\n')

### Question #7: Do the same thing as number 6, but with a while loop this time.

In [16]:
#count = 0

#while count <= 42: # number of columns
#    for i in cat_only.columns:
#        print(f'Frequency table for: {cat_only[i]}')
#        print(cat_only[i].value_counts())
#        print('\n')
#        print(f'Relative frequency table for: {i}')
#        print(round(cat_only[i].value_counts(normalize=True),2))
#        print('\n')
#        print('\n')
#        count = count + 1

### Question #8: Using the Rocket class which we created as part of the week 7 class activity, we are going to create a new class called shuttle. The shuttle class should inherit all of Rocket's characteristics but has one additional parameter: flights_completed which measures the number of flights the shuttle has completed.

Define the shuttle class as the child of the rocket class (please define it in the same jupyter notebook which has all of your other HW answers), and then initialize a new instance of your shuttle class. Once you've created a Shuttle, print the following message: "The shuttle is located at (__, __), and has recorded __ flights", filling the blank spaces to reflect the information for the Shuttle you just created.

In [17]:
class Rocket():
    """Rocket simulates a rocket ship for a game or a physics simulation."""
    
    def __init__(self, x=0, y=0):
        """Each rocket has an (x,y) position in 2D space."""
        self.x = x
        self.y = y
        
    def move_rocket(self, x_increment=0, y_increment=1):
        """Moves the rocket according to the paremeters given.
           Default behavior is to move the rocket up one unit."""
        self.x += x_increment
        self.y += y_increment
        
    def get_distance(self, other_rocket):
        """Calculates the distance from this rocket to another rocket,
           and returns that value"""
        distance = sqrt((self.x-other_rocket.x)**2+(self.y-other_rocket.y)**2)
        return distance

# from class exercise

In [18]:
rocket_0 = Rocket()
rocket_0.x
rocket_0.y
# initializing a rocket, from class exercise

0

In [19]:
class shuttle(Rocket):
    def __init__(self, x=0, y=0):
        super().__init__(x, y) # added shuttle as child of rocket

    def flights_completed(instance):
        instance = 0
        instance = instance + 1
        return instance # making flights completed

In [20]:
shuttle0 = shuttle() # initializing a shuttle
flights = shuttle.flights_completed(0)

print("The shuttle is located at", (shuttle0.x, shuttle0.y), "and has recorded", flights, "flights.")

The shuttle is located at (0, 0) and has recorded 1 flights.


### Question #9: Write a single for loop which randomly creates 11 rockets and 11 shuttles using randomly generated integers. Store this output in two separate lists. Once you have your data, use for loops to print the messages you see below. Note you can use the randint() function from the random module (which you can import with the command from random import randint) to randomly generate the characteristics of your Rockets and Shuttles.

For the shuttles you created, loop through them and print the following messages in your jupyter notebook (Hint: use the enumerate function in your for loop!)

In [21]:
shuttles = []
rockets = []
# creating shuttle and rocket output storage

for i in range(0,11):    
    rockets.append(Rocket(random.randint(0,100), random.randint(0,100)))
    shuttles.append(shuttle(random.randint(0,100), random.randint(0,100)))

In [23]:
shuttles

In [31]:
for index, shuttle in enumerate(shuttles):
    flights = shuttle.flights_completed()
    print("Shuttle " +  str(index) +  " has completed " +  str(flights) +  " flights.")

Shuttle 0 has completed 1 flights.
Shuttle 1 has completed 1 flights.
Shuttle 2 has completed 1 flights.
Shuttle 3 has completed 1 flights.
Shuttle 4 has completed 1 flights.
Shuttle 5 has completed 1 flights.
Shuttle 6 has completed 1 flights.
Shuttle 7 has completed 1 flights.
Shuttle 8 has completed 1 flights.
Shuttle 9 has completed 1 flights.
Shuttle 10 has completed 1 flights.


In [39]:
for index, shuttle in enumerate(shuttles):
    shuttle0 = shuttles[0]
    distance = shuttle0.get_distance(shuttle)
    print("The first shuttle is " +  str(distance) +  " units away from shuttle " +  str(index) + ".")

The first shuttle is 0.0 units away from shuttle 0.
The first shuttle is 54.0 units away from shuttle 1.
The first shuttle is 67.60177512462228 units away from shuttle 2.
The first shuttle is 95.13148795220224 units away from shuttle 3.
The first shuttle is 92.02173656261873 units away from shuttle 4.
The first shuttle is 63.97655820689325 units away from shuttle 5.
The first shuttle is 78.77182237323191 units away from shuttle 6.
The first shuttle is 90.4267659490264 units away from shuttle 7.
The first shuttle is 5.830951894845301 units away from shuttle 8.
The first shuttle is 93.47726996441435 units away from shuttle 9.
The first shuttle is 50.08991914547278 units away from shuttle 10.


In [38]:
for index, rocket in enumerate(rockets):
    distance = shuttle.get_distance(rocket)
    print("The first shuttle is " +  str(distance) +  " units away from rocket " +  str(index) + ".")

The first shuttle is 46.75467891024384 units away from rocket 0.
The first shuttle is 33.83784863137726 units away from rocket 1.
The first shuttle is 39.56008088970496 units away from rocket 2.
The first shuttle is 46.389654018972806 units away from rocket 3.
The first shuttle is 27.459060435491963 units away from rocket 4.
The first shuttle is 38.58756276314948 units away from rocket 5.
The first shuttle is 47.01063709417264 units away from rocket 6.
The first shuttle is 22.847319317591726 units away from rocket 7.
The first shuttle is 44.721359549995796 units away from rocket 8.
The first shuttle is 61.00819617067857 units away from rocket 9.
The first shuttle is 41.97618372363071 units away from rocket 10.


### Bibliography
- YOLO. (2018, March 9). *Answer to the question: How to show all columns' names on a large pandas dataframe?* \[Source code\]. Stack Overflow. https://stackoverflow.com/questions/49188960/how-to-show-all-columns-names-on-a-large-pandas-dataframe YOLO for setting max columns
- Hicks, R. (2026). *week06.ipynb* \[Source Code\]. Modern Data Structures GitHub.
- Hicks, R. (2026). *week07.ipynb* \[Source Code\]. Modern Data Structures GitHub.
- Hicks, R. (2026). *Class Exercise.py* \[Source Code\]. Modern Data Structures GitHub.
- Hicks, R. (2026). *Rocket_class.py* \[Source Code\]. Modern Data Structures GitHub.